### PR TITLE
Fix for 7530 an os 7.20

### PR DESCRIPTION
**Pull request by fcozza**
Hi

I update some string in script

I get updated string from 

http://192.168.0.1:49000/tr64desc.xml

192.168.0.1 is mi FritzBox 7530

regards

### CODE REVIEW

**Comment by jhubig** [comment]
Hey,

thanks. But usually it should still work with FritzOS 7.20. DIdn't it work for you. This is data gathered via UPnP. You can find the description here: https://avm.de/fileadmin/user_upload/Global/Service/Schnittstellen/IGD1.pdf

Here some additional info in the section TR064: https://avm.de/service/schnittstellen/

I hope this helps.

Cheers,
Johannes

**Comment by chr-x** [comment]
Hello,

first, thank you for FritzBoxShell, a nice script!

I tested parts of the suggested PR on FB-7270 (74.06.06) and FB-7590 (154.07.21). Both boxes do fail on `action='GetInfo'` when setting  `location=/igdupnp/control/wandslifconfig1`. Both boxes do succeed when using `location="/upnp/control/wancommonifconfig1"`.

### Test
#### General settings:
```
$  BoxIP=fritz.box BoxUSER=myadmin BoxPW=myadmpass
$ action='GetInfo'
$ uri="urn:dslforum-org:service:WANDSLInterfaceConfig:1"
```


#### Test IGD: ( -> UPnPError )
```
$ location="/igdupnp/control/wandslifconfig1"

### curl command from readout():
$ curl -s -k -m 5 --anyauth -u "$BoxUSER:$BoxPW" "http://$BoxIP:49000$location" -H 'Content-Type: text/xml; charset="utf-8"' -H "SoapAction:$uri#$action" -d "<?xml version='1.0' encoding='utf-8'?><s:Envelope s:encodingStyle='http://schemas.xmlsoap.org/soap/encoding/' xmlns:s='http://schemas.xmlsoap.org/soap/envelope/'><s:Body><u:$action xmlns:u='$uri'></u:$action></s:Body></s:Envelope>"

<?xml version="1.0"?>
 <s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
<s:Body>
<s:Fault>
<faultcode>s:Client</faultcode>
<faultstring>UPnPError</faultstring>
<detail>
<UPnPError xmlns="urn:schemas-upnp-org:control-1-0">
<errorCode>401</errorCode>
<errorDescription>Invalid Action</errorDescription>
</UPnPError>
</detail>
</s:Fault>
</s:Body>
</s:Envelope>
```


#### Test TR-064: ( -> success )
```
$ location=/upnp/control/wandslifconfig1

### curl command from readout():
$ curl -s -k -m 5 --anyauth -u "$BoxUSER:$BoxPW" "http://$BoxIP:49000$location" -H 'Content-Type: text/xml; charset="utf-8"' -H "SoapAction:$uri#$action" -d "<?xml version='1.0' encoding='utf-8'?><s:Envelope s:encodingStyle='http://schemas.xmlsoap.org/soap/encoding/' xmlns:s='http://schemas.xmlsoap.org/soap/envelope/'><s:Body><u:$action xmlns:u='$uri'></u:$action></s:Body></s:Envelope>"

<?xml version="1.0"?>
 <s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
<s:Body>
<u:GetInfoResponse xmlns:u="urn:dslforum-org:service:WANDSLInterfaceConfig:1">
<NewEnable>...</NewEnable>
<NewStatus>...</NewStatus>
<NewDataPath>...</NewDataPath>
<NewUpstreamCurrRate>...</NewUpstreamCurrRate>
<NewDownstreamCurrRate>...</NewDownstreamCurrRate>
<NewUpstreamMaxRate>...</NewUpstreamMaxRate>
<NewDownstreamMaxRate>...</NewDownstreamMaxRate>
<NewUpstreamNoiseMargin>...</NewUpstreamNoiseMargin>
<NewDownstreamNoiseMargin>...</NewDownstreamNoiseMargin>
<NewUpstreamAttenuation>...</NewUpstreamAttenuation>
<NewDownstreamAttenuation>...</NewDownstreamAttenuation>
<NewATURVendor>...</NewATURVendor>
<NewATURCountry>...</NewATURCountry>
<NewUpstreamPower>...</NewUpstreamPower>
<NewDownstreamPower>...</NewDownstreamPower>
</u:GetInfoResponse>
</s:Body>
</s:Envelope>
```


### More Info

- `$ curl -s -m 5 "http://fritz.box:49000/igddesc.xml"` on fb-7590 misses `/igdupnp/control/wandslifconfig1`. (ATM I have no access to fb-7270 for testing.)
- Today [https://avm.de/service/schnittstellen/](url) does not mention IGD. This link has href-attributes that do refer to 47 pdf-files. `$ pdfgrep igdupnp *.pdf` on these files does not return any file that contains matches. `$ pdfgrep upnp/control *.pdf` does return files that match on `/upnp/control`.
- [[IPPF (German)]](https://www.ip-phone-forum.de/threads/traffic-der-fritzbox-6490-cable-%C3%BCber-upnp-abfragen.300243/post-2286024) From this post I read, AVM is switching from IGD-interface to TR-064-interface for years. This results in issues with old code snippets for years.


I suggest to accept the PR. Maybe it is worth to go further and verify the existence of `$location` and `$uri` in `igddesc.xml` and `tr64desc.xml`.

regards

File: fritzBoxShell.sh [hunk 1]
```diff
@@ -235,7 +235,7 @@ LANstate() {
 ### ----------------------------------------------------------------------------------------------------- ###
 
 DSLstate() {
-		location="/igdupnp/control/wandslifconfig1"
+		location="/upnp/control/wandslifconfig1"
```
Comment (jhubig) [inline fritzBoxShell.sh:238]: Accepted

**Review by jhubig — CHANGES_REQUESTED** [review]
I can confirm for the first location update in DSLstate. But for IGDWANstate, IGDDSLLINKstate and IGDIPstate with updated location and uri it dowsn't work for me.

**Comment by chr-x** [comment]
Hello,

Testing fb-7270 (74.06.06) and fb-7590 (154.07.21) with the settings of fcozza:master and jhubig:master deliver the following results. Fb-7590 is connected to WAN via ethernet link, not DSL. This might explain `err2`.

| function     | action    | fb-7270 |             | fb-7590 |              |
|-|-|-|-|-|-|
|                 |              |  (fcozza) | (jhubig) |  (fcozza) | (jhubig) |
| DSLstate() | GetInfo | OK | err1 | OK | err1 |
|||||||
| IGDWANstate() | GetAddonInfos | err1 | err1 | err1 | OK |
|||||||
| IGDDSLLINKstate() | GetDSLLinkInfo       | OK   | err1 | OK   | err2 |
|                   | GetAutoConfig        | OK   | err1 | OK   | err2 |
|                   |GetModulationType     | err1 | err1 | err1 | err2 |
|                   |GetDestinationAddress | OK   | err1 | OK   | err2 |
|                   |GetATMEncapsulation   | OK   | err1 | OK   | err2 |
|                   |GetFCSPreserved       | err1 | err1 | err1 | err2 |
|||||||
| IGDIPstate() | GetConnectionTypeInfo     | OK   | err1 | OK   | OK |
|        | GetAutoDisconnectTime           | err1 | err1 | err1 | OK |
|        | GetIdleDisconnectTime           | err1 | err1 | err1 | OK |
|        | GetStatusInfo                   | OK   | err1 | OK   | OK |
|        | GetNATRSIPStatus                | OK   | err1 | OK   | OK |
|        | GetExternalIPAddress            | OK   | err1 | OK   | OK |
|        | X_AVM_DE_GetExternalIPv6Address | err1 | err1 | err1 | OK |
|        | X_AVM_DE_GetIPv6Prefix          | err1 | err1 | err1 | OK |
|        | X_AVM_DE_GetDNSServer           | err1 | err1 | err1 | OK |
|        | X_AVM_DE_GetIPv6DNSServer       | err1 | err1 | err1 | OK |


OK: valid xml-data
err1: UPnPError 401 (Invalid Action)
err2: UPnPError 403 (Not available Action)


Fb-7270 does not provide IGD-services via '/igdupnp/control':
```
$ curl -s -m 5 http://fritz.box:49000/igddesc.xml | xmlstarlet sel -t -m '//_:controlURL' -v . -n 
/upnp/control/any
/upnp/control/WANCommonIFC1
/upnp/control/WANDSLLinkC1
/upnp/control/WANIPConn1
```

Fb-7590 does provide IGD-services via '/igdupnp/control'.  '/igdupnp/control/wandslifconfig1' is missing:
```
$ curl -s -m 5 http://fritz.box:49000/igddesc.xml | xmlstarlet sel -t -m '//_:controlURL' -v . -n
/igdupnp/control/any
/igdupnp/control/WANCommonIFC1
/igdupnp/control/WANDSLLinkC1
/igdupnp/control/WANIPConn1
/igd2upnp/control/WANIPv6Firewall1

```

With these test results I draw the following conclusions.
The fb-7270 has no AVM-support anymore and should not be in focus.
This PR makes function DSLstate() work.
I cannot comment on IGDDSLLINKstate(). 
IGDWANstate() and IGDIPstate() of the PR do fail. They need rework. Since the original versions do work fine, there is no need to replace them now.

regards

**Comment by jhubig** [comment]
> Hello,
> 
> Testing fb-7270 (74.06.06) and fb-7590 (154.07.21) with the settings of fcozza:master and jhubig:master deliver the following results. Fb-7590 is connected to WAN via ethernet link, not DSL. This might explain `err2`.
> 
> function	action	fb-7270		fb-7590	
> (fcozza)	(jhubig)	(fcozza)	(jhubig)
> DSLstate()	GetInfo	OK	err1	OK	err1
> IGDWANstate()	GetAddonInfos	err1	err1	err1	OK
> IGDDSLLINKstate()	GetDSLLinkInfo	OK	err1	OK	err2
> GetAutoConfig	OK	err1	OK	err2
> GetModulationType	err1	err1	err1	err2
> GetDestinationAddress	OK	err1	OK	err2
> GetATMEncapsulation	OK	err1	OK	err2
> GetFCSPreserved	err1	err1	err1	err2
> IGDIPstate()	GetConnectionTypeInfo	OK	err1	OK	OK
> GetAutoDisconnectTime	err1	err1	err1	OK
> GetIdleDisconnectTime	err1	err1	err1	OK
> GetStatusInfo	OK	err1	OK	OK
> GetNATRSIPStatus	OK	err1	OK	OK
> GetExternalIPAddress	OK	err1	OK	OK
> X_AVM_DE_GetExternalIPv6Address	err1	err1	err1	OK
> X_AVM_DE_GetIPv6Prefix	err1	err1	err1	OK
> X_AVM_DE_GetDNSServer	err1	err1	err1	OK
> X_AVM_DE_GetIPv6DNSServer	err1	err1	err1	OK
> OK: valid xml-data
> err1: UPnPError 401 (Invalid Action)
> err2: UPnPError 403 (Not available Action)
> 
> Fb-7270 does not provide IGD-services via '/igdupnp/control':
> 
> ```
> $ curl -s -m 5 http://fritz.box:49000/igddesc.xml | xmlstarlet sel -t -m '//_:controlURL' -v . -n 
> /upnp/control/any
> /upnp/control/WANCommonIFC1
> /upnp/control/WANDSLLinkC1
> /upnp/control/WANIPConn1
> ```
> 
> Fb-7590 does provide IGD-services via '/igdupnp/control'. '/igdupnp/control/wandslifconfig1' is missing:
> 
> ```
> $ curl -s -m 5 http://fritz.box:49000/igddesc.xml | xmlstarlet sel -t -m '//_:controlURL' -v . -n
> /igdupnp/control/any
> /igdupnp/control/WANCommonIFC1
> /igdupnp/control/WANDSLLinkC1
> /igdupnp/control/WANIPConn1
> /igd2upnp/control/WANIPv6Firewall1
> ```
> 
> With these test results I draw the following conclusions.
> The fb-7270 has no AVM-support anymore and should not be in focus.
> This PR makes function DSLstate() work.
> I cannot comment on IGDDSLLINKstate().
> IGDWANstate() and IGDIPstate() of the PR do fail. They need rework. Since the original versions do work fine, there is no need to replace them now.
> 
> regards

Wow. I call this an exhaustive test report. Thanks for that. Really appreciated.
Just to know if I understood it correctly. @chr-x : Please correct me if I'm wrong:

Only the DSLstate needs to be reworked currently? If yes, this is what I asked to rework. I will wait to give @fcozza time to check the change request. If needed, the I will do the change.

Cheers,
Johannes

**Comment by chr-x** [comment]
> Just to know if I understood it correctly. @chr-x : Please correct me if I'm wrong:
> 
> Only the DSLstate needs to be reworked currently? If yes, this is what I asked to rework. I will wait to give @fcozza time to check the change request. If needed, the I will do the change.

Sorry for being ambiguous.
I suggest to make use of ```location="/upnp/control/wandslifconfig1"``` (line 138 of ce79d39) in the upstream repo [jhubig:master](/jhubig/fritzBoxShell/tree/master). In my sight it is just that related to this pull request, not more.

A wish for the future. FritzBoxShell should make use of the output of ```igddesc.xml``` and ```tr64desc.xml``` for setting ```uri=``` and ```location=``` and for verifying the existence of ```action=```. I guess this wish is out of the scope of this pull request and if desired it should go to a new issue or pull request.

@jhubig: I hope my sentences here do improve our understanding. May be I misunderstand what you expect from 'reworking' function DSLstate() by saying: '_Only the DSLstate needs to be reworked..._'. In my view DSLstate() is the only function in this pull request, that is working fine and improving fritzBoxShell at jhubig:master. I think, it can be used unmodified.

regards,
Christian

**Comment by jhubig** [comment]
Hello,

I hope fine for everyone:

@fcozza: I've closed this pull request and just changed the DLSstate() function via commit https://github.com/jhubig/FritzBoxShell/commit/6eea414f727ed3e03cc18ac95b4fc71586323482
@chr-x : I've created a new issue for your future wish 😀 : #17 

Cheers,
Johannes

**Pull request closed without merging**